### ISSUE TITLE
big-code-analysis: read_file_with_eol swallows probe read_exact I/O errors as Ok(None)

### ISSUE THREAD

**Issue by dekobon** [dekobon/big-code-analysis#804]
## Summary

In `read_file_with_eol`, a failed `read_exact` of the UTF-8 probe window
is unconditionally mapped to `Ok(None)`, swallowing genuine I/O errors
and contradicting the function's documented error contract.

## Location

- `src/tools.rs:104-119` (`read_file_with_eol`)

## Evidence

```rust
let mut file = File::open(path)?;

let probe_len = UTF8_PROBE_BYTES.min(file_size);
let mut start = vec![0; probe_len];
let start = if file.read_exact(&mut start).is_ok() {
    ... BOM handling ...
} else {
    return Ok(None);   // <-- ALL read_exact errors become Ok(None)
};
```

The doc comment (lines 82-85) states:

> Returns any [`std::io::Error`] surfaced by [`File::open`] ... or by
> the subsequent reads from the open file handle. A non-UTF-8 head or a
> too-small file is reported via `Ok(None)`, not an error.

But the *first* subsequent read (`read_exact`) does not surface its
error: every failure — not just `UnexpectedEof` from a short read — is
collapsed to `Ok(None)`. `read_exact` can fail with `Interrupted`,
permission errors, or hardware/`io::ErrorKind::Other` mid-read, all of
which here masquerade as "this file is too small / not UTF-8."

This also conflates two distinct conditions: a file that genuinely
shrank below `probe_len` after the `fs::metadata` call (`UnexpectedEof`)
versus a real I/O fault. Only the latter half of the read path
(`file.read_to_end(&mut data)?` at line 146) actually propagates errors
as the doc promises.

## Expected Behavior

A genuine I/O error during the probe read should propagate as `Err`
(matching the documented contract and the `read_to_end` call below it).
Only a clean short-read / EOF should map to `Ok(None)`.

## Actual Behavior

Every `read_exact` failure, including real I/O errors, returns
`Ok(None)`, silently dropping the file from analysis with no error
surfaced to the caller.

## Impact

A transient or hard I/O error on the first 64 bytes of a file is
indistinguishable from "small/binary file" — the file is silently
skipped instead of reported. Callers (`dispatch.rs`, the Python
`analyze_path`, which rely on `Err` for I/O reporting per
`big-code-analysis-py/src/analysis.rs:366`) cannot tell a read failed.

---

**Resolution:** Fixed in `872e7cd4` (branch `fix/issue-803`). The probe `read_exact` now distinguishes a clean short read (`UnexpectedEof` -> `Ok(None)`) from any other error kind, which propagates as `Err`. Regression test `read_eol_propagates_non_eof_io_error` added in `src/tools_tests.rs`.

**Comment by dekobon** [comment]
## General Plan for Fix

### Root Cause

`read_exact`'s error is tested with `.is_ok()` and any falsy result is
folded into `Ok(None)`, so genuine I/O errors are indistinguishable from
a clean short read.

### Implementation Steps

1. Replace the `is_ok()`/`else return Ok(None)` shape with an explicit
   match on the `read_exact` result:
   - `Ok(())` → proceed with BOM handling.
   - `Err(e) if e.kind() == io::ErrorKind::UnexpectedEof` → `Ok(None)`
     (file shrank below `probe_len` after the metadata call; the
     "too-small file" case the doc describes).
   - `Err(e)` → `Err(e)` (propagate, matching the doc and the
     `read_to_end(&mut data)?` below).
2. No doc change needed if the behavior is fixed to match the existing
   comment; otherwise align the comment.

### Tests to Add/Update

- A short-read scenario (file truncated to < probe_len, or a file whose
  metadata over-reports size) still yields `Ok(None)`.
- Hard to inject a non-EOF I/O error in a unit test portably; at minimum
  assert the `UnexpectedEof` branch returns `Ok(None)` and document the
  propagation path for other errors. Consider a seam (e.g. reading from a
  `/proc`-style or permission-restricted fd) if a portable reproducer is
  available.

### Verification

- [ ] `cargo test -p big-code-analysis` passes
- [ ] `cargo clippy -p big-code-analysis` clean
- [ ] Manual: a readable small file (<= probe_len real bytes) is skipped
      as before; no regression to the EOF path

**Comment by dekobon** [comment]
Fixed in `872e7cd4` on branch `fix/issue-803`.

The probe `read_exact` in `read_file_with_eol` (`src/tools.rs`) no
longer collapses every failure to `Ok(None)`. It now matches on the
error kind: a clean short read (`io::ErrorKind::UnexpectedEof`, the
file shrank below the probe between `metadata` and the read) yields
`Ok(None)`; any other error kind propagates as `Err`, matching the
documented contract that real I/O faults surface as errors.
`Interrupted` is not special-cased — `read_exact` already retries it
internally.

Regression test in `src/tools_tests.rs`:
- `read_eol_propagates_non_eof_io_error` (`#[cfg(unix)]`) — opening a
  directory succeeds but the probe read fails with `IsADirectory`
  (non-EOF), so the call must return `Err`. Confirmed to fail against
  the pre-fix collapse-all code (test-via-revert).
